persistence/src/main/scala/akka/persistence/kafka/journal/KafkaJournal.scala

```diff
@@ -27,9 +28,19 @@ class KafkaJournal(config: Config) extends AsyncWriteJournal { actor =>
 
   implicit val system       : ActorSystem              = context.system
   implicit val executor     : ExecutionContextExecutor = context.dispatcher
-  implicit val contextShift : ContextShift[IO]         = IO.contextShift(executor)
-  implicit val parallel     : Parallel[IO]             = IO.ioParallel(contextShift)
-  implicit val timer        : Timer[IO]                = IO.timer(executor)
+
+  private val (blocking, blockingShutdown)   = IORuntime.createDefaultBlockingExecutionContext()
+  private val (scheduler, schedulerShutdown) = IORuntime.createDefaultScheduler()
```

Comment:
I'd prefer reusing thread pools, could you please at least create separate task about reusing scheduler from Akka.

Reply:
The main compute pool is reused here (`executor`). As for the blocking pool (`Blocker`), I haven't seen any custom definitions of that, which means it was using the default cached one (`BlockerPlatform#apply`) in which case it's no different from what we use here.
As for the scheduler, it works a bit differently in CE3. Previously, `IO.timer(ec)` was used where `ec` was the thread pool to execute tasks after schedule and actual scheduling was performed on a global `ScheduledExecutorService` (`IOTimer#apply`). So we didn't specify a custom `ScheduledExecutorService`, only the thread pool for binding continuations (`ec`). Now we don't need this `ec` to bind continuations and the scheduler itself is still a default one (as before).

Reply:
> As for the blocking pool (Blocker), I haven't seen any custom definitions of that, which means it was using the default cached one (BlockerPlatform#apply) in which case it's no different from what we use here.

Afaik we used custom resizable pool 

```
Executors.blocking
```

Could we keep it used ?

Reply:
I couldn't find anything in the current `master` branch by searching for `Blocker` and `Executors.blocking` is used only in a couple of places (and `KafkaJournal` isn't one of them [link](https://github.com/evolution-gaming/kafka-journal/search?q=Executors.blocking)) and I haven't changed that, so I'm not sure what you mean.

I can use `Executors.blocking` here, but there's a catch - it returns `Resource` which we need to allocate and for that we need `IORuntime` which is basically created with the result of that Resource 😄 
I can do something like this here, if it's OK with you:
```scala
  val (blocking, blockingShutdown) = 
    Executors.blocking[IO]("kafka-journal-blocking")
      .allocated
      .unsafeRunSync()(cats.effect.unsafe.IORuntime.global)
  implicit val ioRuntime: IORuntime = IORuntime(
    compute = executor,
    blocking = blocking,
    scheduler = scheduler,
    shutdown = () => {
      blockingShutdown.unsafeRunSync()(cats.effect.unsafe.IORuntime.global)
      schedulerShutdown()
    },
    config = IORuntimeConfig()
  )
```

However, that's not very clean and I'm not sure if there's much of a difference, considering that it has the same implementation as `IORuntime.createDefaultBlockingExecutionContext()`:

Source:
[IORuntime.createDefaultBlockingExecutionContext](https://github.com/typelevel/cats-effect/blob/27307cfe6cd9ea3cb230a8f9be5e16f624a914cd/core/jvm/src/main/scala/cats/effect/unsafe/IORuntimeCompanionPlatform.scala#L106-L116) where `Executors.newCachedThreadPool` has the following implementation:
```java
public static ExecutorService newCachedThreadPool(ThreadFactory threadFactory) {
        return new ThreadPoolExecutor(0, Integer.MAX_VALUE,
                                      60L, TimeUnit.SECONDS,
                                      new SynchronousQueue<Runnable>(),
                                      threadFactory);
    }
```

[Executors.blocking](https://github.com/evolution-gaming/kafka-journal/blob/14b9916782ab05a3e9da1199aac4958f65e5409e/journal/src/main/scala/com/evolutiongaming/kafka/journal/util/Executors.scala#L14-L23) where `ThreadPoolOf[F](2, Int.MaxValue, threadFactory)` does just [the same](https://github.com/evolution-gaming/kafka-journal/blob/14b9916782ab05a3e9da1199aac4958f65e5409e/journal/src/main/scala/com/evolutiongaming/kafka/journal/execution/ThreadPoolOf.scala) (except for `minSize` which is 2 instead of 0 and also CE version spawns daemonic threads) and it's also not configurable in terms of size, you can only pass a thread prefix, just like you can pass it  to `IORuntime.createDefaultBlockingExecutionContext`.

Reply:
I believe it's a common approach in cats-effect to use an unbounded cached pool for blocking operations. Even in cats-effect 2.x.x, if we take a look at the implementation which is summoned when using `Blocker[IO]` it uses the same unbounded `Executors.newCachedThreadPool` [link](https://github.com/typelevel/cats-effect/blob/series/2.x/core/jvm/src/main/scala/cats/effect/internals/BlockerPlatform.scala#L31-L38)

Reply:
I'm happy to see things evolving this way!

It used to be a different opinion of community, code in subject existed before `Blocker`… and even before we started to adopt cats & cats-effect :)